### PR TITLE
ceph: Add holder pod if csi host networking is disabled

### PR DESCRIPTION
**Pull request by Madhu-1**
**Description of your changes:**

If csi is configured not the use the hostnetworking, deploy the holder pod for executing the commands with nsenter.
The implementation is the same as multus.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

Note:
```
Added DNM label as testing is still pending
```

### CODE REVIEW

File: pkg/operator/ceph/csi/spec.go
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go:390]: How about another var that helps with readability? Otherwise the conditions are hard to understand.
```suggestion
	holderEnabled := !CSIParam.EnableCSIHostNetwork || multusApplied
	if (holderEnabled) && !v.SupportsNsenter() {
```

File: pkg/operator/ceph/csi/spec.go
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go:485]:
```suggestion
		if holderEnabled {
```

File: pkg/operator/ceph/csi/spec.go
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go:752]: Are these clusters only the clusters configured with the holder pod? This loops makes it look like all clusters.
```suggestion
	for _, cluster := range r.holderPodClusters {
```

File: .github/workflows/canary-integration-test.yml
Comment (travisn) [inline .github/workflows/canary-integration-test.yml:1211]: Instead of adding a new test, what if we just change existing tests to run without host networking? We want to make this the default anyway. And do we already have a test that restarts the CSI plugins so we could verify that it's working?

File: pkg/operator/ceph/csi/controller.go
Comment (travisn) [inline pkg/operator/ceph/csi/controller.go:172]:
```suggestion
		if holderEnabled {
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/spec.go
Comment (travisn) [reply pkg/operator/ceph/csi/spec.go:752]: Actually, I think we just need to have all clusters in this slice, so we should keep it as `r.clusters`. For the telemetry in #10341 I need the clusterInfo for all clustesr to properly report the csi version to all clusters. However, this is really making me think we need to refactor how the clusters interact with the csi driver. We need to find a way to decouple the csi configuration and cluster reconcile, so that they won't be blocked on each other and the needed configmaps or other resources can be updated as soon as the cluster and csi driver are configured. The problem with the current design is that the csi driver won't even start until the clusterInfo is available for all clusters. It's fine if there is a single cluster, but not if there is more than one cluster since it could take minutes before the csi driver would start in that case.

Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go:752]: If the host networking is disabled we need to deploy a holder pod for all the ceph clusters. so we need all cluster details. sounds good to refactor, maybe we can think about it as a follow-up?

**Comment by Madhu-1** [comment]
working as expected and even CI is passing. @travisn @leseb PTAL

File: .github/workflows/canary-integration-test.yml
Comment (travisn) [reply .github/workflows/canary-integration-test.yml:1211]: Sure we can keep it, that's fine. I was just asking since it's a general pattern to use the existing test suites and enable different settings for the different suites, instead of creating new tests for every setting.

File: pkg/operator/ceph/csi/controller.go
Comment (travisn) [inline pkg/operator/ceph/csi/controller.go:189]:
```suggestion
			logger.Debugf("not a multus cluster %q or CSI_ENABLE_HOST_NETWORK is true, not deploying the ceph-csi plugin holder", request.NamespacedName)
```

File: pkg/operator/ceph/csi/spec.go
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go:374]: Looks like all the usages of `multusApplied` should just be replaced by the new holderEnabled check, right? So it seems we don't need the `multusApplied` var anymore, and consolidate line 390 here.
```suggestion
	holderEnabled := !CSIParam.EnableCSIHostNetwork
```

File: pkg/operator/ceph/csi/spec.go
Comment (travisn) [reply pkg/operator/ceph/csi/spec.go:752]: Agreed, as a separate PR sounds good, it's going to be a bigger change.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/spec.go
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go:374]: We still need this one as we don't care about the value of `CSIParam.EnableCSIHostNetwork` if its a multus enabled ceph cluster. As the default value of `CSIParam.EnableCSIHostNetwork` is true and ceph cluster is multus enabled. this helps to identify whether to create a holder pod or not.

File: .github/workflows/canary-integration-test.yml
Comment (Madhu-1) [reply .github/workflows/canary-integration-test.yml:1211]: Once pod networking becomes stable will remove this one as the default will be pod networking and also I will take a look at it later how to test both pod network and host network in the same test suite by changing multiple settings instead of creating new tests for every setting.

File: tests/scripts/validate_cluster.sh
Comment (leseb) [inline tests/scripts/validate_cluster.sh:105]:
```suggestion
    if [ -n $IS_MULTUS ]; then
```

**Review by leseb — CHANGES_REQUESTED** [review]
nit

**Comment by leseb** [comment]
> > Nit. Also do we need to update the CSI templates?
> 
> @leseb For which parameters? i didn't see any problem with CI tests and local testing. let me know if i missed anything.

For hostnetworking here: https://github.com/rook/rook/blob/master/pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml#L25

Do we need to flip https://github.com/rook/rook/blob/master/pkg/operator/ceph/csi/csi.go#L87 to false?

**Comment by leseb** [comment]
> 



> > > > Nit. Also do we need to update the CSI templates?
> > > 
> > > 
> > > @leseb For which parameters? i didn't see any problem with CI tests and local testing. let me know if i missed anything.
> > 
> > 
> > For hostnetworking here: https://github.com/rook/rook/blob/master/pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml#L25
> 
> As the `CSI_ENABLE_HOST_NETWORK` is already set to false in configmap/ENV cephcsi pods will use pod networking only.
> 
> > Do we need to flip https://github.com/rook/rook/blob/master/pkg/operator/ceph/csi/csi.go#L87 to false?
> 
> lets change the default later once its stable

Isn't the next Rook release sufficient to give it time to stabilize?

**Comment by Madhu-1** [comment]
> > 
> 
> > > > > Nit. Also do we need to update the CSI templates?
> > > > 
> > > > 
> > > > @leseb For which parameters? i didn't see any problem with CI tests and local testing. let me know if i missed anything.
> > > 
> > > 
> > > For hostnetworking here: https://github.com/rook/rook/blob/master/pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml#L25
> > 
> > 
> > As the `CSI_ENABLE_HOST_NETWORK` is already set to false in configmap/ENV cephcsi pods will use pod networking only.
> > > Do we need to flip https://github.com/rook/rook/blob/master/pkg/operator/ceph/csi/csi.go#L87 to false?
> > 
> > 
> > lets change the default later once its stable
> 
> Isn't the next Rook release sufficient to give it time to stabilize?

let's backport this to 1.9 and change the default to pod network in master and make it default in 1.10 @leseb WDYT?

**Review by leseb — APPROVED** [review]

File: tests/scripts/validate_cluster.sh
Comment (Madhu-1) [inline tests/scripts/validate_cluster.sh:105]: @leseb looks like with this change CI is failing.

Comment (Madhu-1) [reply tests/scripts/validate_cluster.sh:105]: yes not sure why this fails, shall i revert back it back as the functionality remains same?

Comment (Madhu-1) [reply tests/scripts/validate_cluster.sh]: pushing it again to check `! -z` behaviour

Comment (leseb) [reply tests/scripts/validate_cluster.sh]: Are we sure it's that change that makes the CI fail?

Comment (leseb) [reply tests/scripts/validate_cluster.sh]: Looks like it's failing again so `-n` was correct.

Comment (Madhu-1) [reply tests/scripts/validate_cluster.sh]: The failure is something else. restarted to check it.

File: pkg/operator/ceph/csi/controller.go
Comment (travisn) [inline pkg/operator/ceph/csi/controller.go:77]: For now these are only the clusters that use the holder pod, right? In the future refactor PR, then we can rename it to `clusters` if we collect all clusters.
```suggestion
		holderClusters:         []ClusterDetail{},
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/spec.go
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go:374]: Yes correct, removed extra variable

File: pkg/operator/ceph/csi/controller.go
Comment (leseb) [inline pkg/operator/ceph/csi/controller.go:54]: In the end, they are `ClustersDetail` or something. The `holderClusters` is confusing, `multusClusters` served a real purpose since it was the only scenario that would enable the holder pod. Now, we need to make this more generic since it"ll become the default really soon.

**Review by leseb — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/controller.go
Comment (Madhu-1) [reply pkg/operator/ceph/csi/controller.go:54]: @leseb made this change as requested by  @travisn here https://github.com/rook/rook/pull/10342#discussion_r887282610. the suggestion is to move back to clusters?

Comment (leseb) [reply pkg/operator/ceph/csi/controller.go:54]: Alright, let's discuss that with @travisn :)

**Comment by Rakshith-R** [comment]
let's keep nfsplugin on hostnetworking by default until https://github.com/rook/rook/issues/10372 is resolved ? 
Otherwise, It would face the same troubles we faced when we didn't have holder pod design for rbd and cephfs.

**Comment by Rakshith-R** [comment]
cc @nixpanic

**Comment by Rakshith-R** [comment]
cc @nixpanic  @BlaineEXE

**Comment by Madhu-1** [comment]
> let's keep nfsplugin on hostnetworking by default until #10372 is resolved ? Otherwise, It would face the same troubles we faced when we didn't have holder pod design for rbd and cephfs.

Yes, this PR is not touching the NFS driver. I created issue #10372 to track pod networking for NFS

File: pkg/operator/ceph/csi/controller.go
Comment (travisn) [reply pkg/operator/ceph/csi/controller.go:54]: We could just call it `clusters`, but then I think all clusters should be added to the list. If it's a subset of clusters only for holders, then `holderClusters` makes sense to me. The main blocker to adding all clusters is just what Seb and I discussed offline where it is a problem if we block creating the csi driver until all clusters are initialized. It seems we should allow the csi driver to be started independent of whether the clusters are initialized, then the holder pod initialization would just need to wait for cluster initialization.

Comment (Madhu-1) [reply pkg/operator/ceph/csi/controller.go:54]: @travisn @leseb what's the conclusion on this one, should we keep it `clusters` or `holderClusters`?

Comment (leseb) [reply pkg/operator/ceph/csi/controller.go:54]: So I've made a PR for the blocking issue: https://github.com/rook/rook/pull/10380. I think `clusters` makes sense.

Comment (travisn) [reply pkg/operator/ceph/csi/controller.go:54]: but it's not all clusters, right? so we need a more specific name than `clusters` IMO.

Comment (Madhu-1) [reply pkg/operator/ceph/csi/controller.go]: Done revert back to clusters now.

Comment (Madhu-1) [reply pkg/operator/ceph/csi/controller.go]: I did change it back in parallel to @travisn comments. once we decide on this one i will update it.

File: tests/scripts/validate_cluster.sh
Comment (leseb) [reply tests/scripts/validate_cluster.sh:105]: Use `-n` now?

Comment (Madhu-1) [reply tests/scripts/validate_cluster.sh]: @leseb with `-n` its failing again CI logs here https://github.com/rook/rook/runs/6750483568?check_suite_focus=true. not sure why it's failing

File: pkg/operator/ceph/csi/controller.go
Comment (leseb) [reply pkg/operator/ceph/csi/controller.go]: @travisn and I just discussed and agreed on `clustersWithHolder`, @Madhu-1 wdyt?

Comment (Madhu-1) [reply pkg/operator/ceph/csi/controller.go]: looks good to me.

**Review by travisn — COMMENTED** [review]
LGTM after the CI issue is fixed

**Review by leseb — COMMENTED** [review]
@Madhu-1 looks like the CI is failing but I don't think it's a CI issue anymore.

**Comment by Madhu-1** [comment]
> @Madhu-1 looks like the CI is failing but I don't think it's a CI issue anymore.

Yes, something is wrong, csi holder pod is not getting deployed. will check and update

**Review by travisn — APPROVED** [review]

**Pull request merged**
